### ISSUE TITLE
Find all duplicate strings 

### ISSUE THREAD

**Issue by peter-michalski** [JacquesCarette/Drasil#2832]
Find all duplicate strings in the source code.

We will want to minimize the number of times the same string occurs. 

Leading or trailing spaces may be considered the same.

**Comment by peter-michalski** [comment]
[output_strings2.txt](https://github.com/JacquesCarette/Drasil/files/7158548/output_strings2.txt)
[output_final2.txt](https://github.com/JacquesCarette/Drasil/files/7158553/output_final2.txt)

output_strings2.txt: Strings of 5 characters or longer (strings that appear 3+ times in the source code)
output_final2.txt: Files that contain the strings

**Comment by smiths** [comment]
Thank you @peter-michalski.  Good start.

The list of files seems broader then I would have expected.  You've included files other than the source code, like `.git/index`, Dan's thesis, presentation files, etc.  We can restrict the exercise to the code files.  The files are in the `code` folder, but not in `stable`, since `stable` is generated files.  The code files will have the extension `.hs`.

It would be nice to have the list of strings sorted by the number of appearances, so that we can start looking at the ones that are repeated the most first.  

If you wrote a script to do your analysis, it would be good to share the script.

**Comment by peter-michalski** [comment]
Updated files:
[output_strings3.txt](https://github.com/JacquesCarette/Drasil/files/7162996/output_strings3.txt)
[output_final3.txt](https://github.com/JacquesCarette/Drasil/files/7162997/output_final3.txt)

script file getstrings3.sh (cannot upload .sh):

#!/bin/bash  

grep -rihEo --exclude-dir=./Drasil/code/stable --include \*.hs '".{5,100}"' ./Drasil/code | awk '++A[$0]==3,1000' > output_strings3.txt

while read -r  p; do
    echo $p >>output_final3.txt
    stringinstances=$(grep -rihEo --exclude-dir=./Drasil/code/stable --include \*.hs "$p" ./Drasil/code | wc -l)
    echo "Number of instances of string: ${stringinstances}">>output_final3.txt
    filescontainstring=$(grep -riEo --exclude-dir=./Drasil/code/stable --include \*.hs "$p" ./Drasil/code | sort -u | wc -l)    
    echo "Number of files containing the string: ${filescontainstring}" >>output_final3.txt
    grep -riEo --exclude-dir=./Drasil/code/stable --include \*.hs "$p" ./Drasil/code | sort -u >>output_final3.txt
    printf '%s\n'>>output_final3.txt
done <output_strings3.txt

**Comment by smiths** [comment]
Looks great @peter-michalski.  I think you've provided a good starting point for refactoring some of the code to remove some of these repeats.  @JacquesCarette will likely have suggestions on the best place to start.  :smile:.

This task is done, but lets hold off closing this issue until @JacquesCarette and @balacij have had a chance to look at the results.

**Comment by JacquesCarette** [comment]
This is very nice indeed. Still needs a little bit of tweaking though, as things like
```
"This", phrase equation, S "is based on the"
```
are in the `output_strings3.txt`, where this is really 2 strings.  In other words, the strings should not contain `"`.

Nevertheless, `output_final3.txt` contains a lot of actionable material. Way too much, really.  I'm rather tempted to jump right in and "fix" them... but I won't.

Also, I noticed things like
```
"(expr1)"
Number of instances of string: 0
Number of files containing the string: 0
```
which are also odd.

@peter-michalski if you could investigate/fix those oddities please?

Also, in parallel, I think it's worth starting to investigate all the cases that occur 9 or more times (I took a look through everything, and that seemed like a reasonable first cut off). It's probably worth creating one issue per instance. Yes, I know, that will be a lot of issues. But I think there will be sufficient diversity in resolutions so as to make it worth it.

Some of them are going to be fun to discuss. "where" and "using" leapt out at me as being two such.

**Comment by peter-michalski** [comment]
Updated output:

[duplicate_strings.txt](https://github.com/JacquesCarette/Drasil/files/7180205/duplicate_strings.txt)
[duplicate_string_summary.txt](https://github.com/JacquesCarette/Drasil/files/7180206/duplicate_string_summary.txt)

I will upload the .sh scripts to the repo

**Comment by peter-michalski** [comment]
@JacquesCarette can we close this issue?

**Comment by JacquesCarette** [comment]
Yes.
